pkg/pmem-csi-operator/controller/deployment/controller_driver.go

```diff
@@ -194,8 +175,21 @@ func (d *PmemCSIDriver) reconcileDeploymentChanges(r *ReconcileDeployment, exist
 	return requeue, err
 }
 
-func (d *PmemCSIDriver) initDeploymentSecrests(r *ReconcileDeployment) error {
+func (d *PmemCSIDriver) deployObjects(r *ReconcileDeployment) error {
+	objects, err := d.getDeploymentObjects(r)
+	if err != nil {
+		return err
+	}
+	for _, obj := range objects {
+		if err := r.Create(obj); err != nil {
+			return err
+		}
+	}
+	return nil
+}
 
+// getDeploymentObjects returns all objects that are part of a driver deployment.
+func (d *PmemCSIDriver) getDeploymentObjects(r *ReconcileDeployment) ([]runtime.Object, error) {
```

Comment:
@pohly By adding secrets creation to `getDeploymentObjects`, calling this method more than once creates new keys/certificates for every call in case the deployment chooses to use defaults certificates.

Reply:
Is that a bad thing? It would ensure that those certificates are always fresh.

We haven't sorted out how to handle periodic updates of those certificates and the current code also doesn't have your "populate deployment from objects" code yet. Let's figure out how to avoid updating the secrets too often later, okay?

Reply:
> Is that a bad thing? It would ensure that those certificates are always fresh.
> We haven't sorted out how to handle periodic updates of those certificates and the current code also doesn't have your "populate deployment from objects" code yet. Let's figure out how to avoid updating the secrets too often later, okay?

That depends on how often we call this method, At least I could think of on every restart of operator pod. Updating secrets also lead to the restart of all driver pods.

Reply:
> That depends on how often we call this method, At least I could think of on every restart of operator pod.

Which isn't very often. And even if it was called more often, why would that be a reason to be concerned?

> Updating secrets also lead to the restart of all driver pods.

No, it doesn't: https://kubernetes.io/docs/concepts/configuration/secret/#mounted-secrets-are-updated-automatically

It's still not perfect because the certificates may get rolled out on different nodes by kubelet at different times, which leads to a time window where controller and node cannot communicate. When designing the certificate update feature we need to take that into account, for example by keeping old and new CA root as trusted for a certain period of time.

For now I think it is okay that the system eventually recovers.

Reply:
> Which isn't very often. And even if it was called more often, why would that be a reason to be concerned?
> 
Ok, If you feel it's not an issue.
 
> No, it doesn't: https://kubernetes.io/docs/concepts/configuration/secret/#mounted-secrets-are-updated-automatically
> It's still not perfect because the certificates may get rolled out on different nodes by kubelet at different times, which leads to a time window where controller and node cannot communicate. When designing the certificate update feature we need to take that into account, for example by keeping old and new CA root as trusted for a certain period of time.
> For now I think it is okay that the system eventually recovers.

Interesting, Thanks for sharing.